src/ol/proj.js

```diff
@@ -463,11 +473,55 @@ export function getTransformFromProjections(
 ) {
   const sourceCode = sourceProjection.getCode();
   const destinationCode = destinationProjection.getCode();
-  let transformFunc = getTransformFunc(sourceCode, destinationCode);
-  if (!transformFunc) {
-    transformFunc = identityTransform;
+  const transformFunc = getTransformFunc(sourceCode, destinationCode);
+  const sourceMatrix = sourceProjection.getMatrix();
+  const destinationMatrix = destinationProjection.getMatrix();
+  if (!sourceMatrix && !destinationMatrix) {
+    return transformFunc || identityTransform;
+  }
+
+  let sourceTransform, destinationTransform;
+  if (sourceMatrix) {
+    const matrix = invertMatrix(sourceMatrix.slice());
+    sourceTransform = createTransformFromCoordinateTransform(
+      function (coordinate) {
+        return applyMatrix(matrix, coordinate);
+      },
+    );
   }
-  return transformFunc;
+  if (destinationMatrix) {
+    const matrix = destinationMatrix;
+    destinationTransform = createTransformFromCoordinateTransform(
+      function (coordinate) {
+        return applyMatrix(matrix, coordinate);
+      },
+    );
+  }
+  return (
```

Comment:
A single transform could be used directly here, but any performance gain from the extra code would be minimal
```
  if ((!sourceTransform || !destinationTransform) && !transformFunc) {
    return sourceTransform || destinationTransform;
  }
```